src/Containers/DenseAxisArray.jl

```diff
@@ -141,62 +227,37 @@ end
 
 Base.eachindex(A::DenseAxisArray) = CartesianIndices(size(A.data))
 
-lookup_index(i::Colon, lookup) = Colon()
-lookup_index(i, lookup) = lookup[i]
-
-# Lisp-y tuple recursion trick to handle indexing in a nice type-
-# stable way. The idea here is that `_to_index_tuple(idx, lookup)`
-# performs a lookup on the first element of `idx` and `lookup`,
-# then recurses using the remaining elements of both tuples.
-# The compiler knows the lengths and types of each tuple, so
-# all of the types are inferable.
-function _to_index_tuple(idx::Tuple, lookup::Tuple)
-    return tuple(
-        lookup_index(first(idx), first(lookup)),
-        _to_index_tuple(Base.tail(idx), Base.tail(lookup))...,
-    )
-end
-
-# Handle the base case when we have more indices than lookups:
-function _to_index_tuple(idx::NTuple{N}, ::NTuple{0}) where {N}
-    return ntuple(k -> begin
-        i = idx[k]
-        (i == 1) ? 1 : error("invalid index $i")
-    end, Val(N))
+function Base.to_index(A::DenseAxisArray{T,N}, idx) where {T,N}
+    if length(idx) < N
+        throw(BoundsError(A, idx))
+    elseif any(i -> !isone(idx[i]), (N+1):length(idx))
+        throw(KeyError(idx))
+    end
+    return ntuple(Val(N)) do k
+        return getindex(A.lookup[k], idx[k])
+    end
 end
 
-# Handle the base case when we have fewer indices than lookups:
-_to_index_tuple(idx::NTuple{0}, lookup::Tuple) = ()
-
-# Resolve ambiguity with the above two base cases
-_to_index_tuple(idx::NTuple{0}, lookup::NTuple{0}) = ()
+_is_range(::Any) = false
+_is_range(::Union{Vector{Int},Colon}) = true
 
-to_index(A::DenseAxisArray, idx...) = _to_index_tuple(idx, A.lookup)
-
-# Doing `Colon() in idx` is relatively slow because it involves
-# a non-unrolled loop through the `idx` tuple which may be of
-# varying element type. Another lisp-y recursion trick fixes that
-has_colon(idx::Tuple{}) = false
-has_colon(idx::Tuple) = isa(first(idx), Colon) || has_colon(Base.tail(idx))
-
-# TODO: better error (or just handle correctly) when user tries to index with a range like a:b
-# The only kind of slicing we support is dropping a dimension with colons
 function Base.getindex(A::DenseAxisArray{T,N}, idx...) where {T,N}
-    length(idx) < N && throw(BoundsError(A, idx))
-    if has_colon(idx)
-        DenseAxisArray(
-            A.data[to_index(A, idx...)...],
-            (ax for (i, ax) in enumerate(A.axes) if idx[i] == Colon())...,
-        )
-    else
-        return A.data[to_index(A, idx...)...]
+    new_indices = Base.to_index(A, idx)
+    new_data = A.data[new_indices...]
+    if !any(_is_range, new_indices)
+        return new_data::T
     end
+    new_axes = (
+        axis[index] for
+        (axis, index) in zip(A.axes, new_indices) if _is_range(index)
```

Comment:
Is this type-stable ? There should be a way to make this type-stable with `ntuple` or with a recursion.

Reply:
On Julia <=1.5, slicing does not infer properly, but then neither does the current `Colon` approach.

However, on Julia 1.6, this does infer, and there is a `@inferred` test to check that.

Reply:
Fixed. Thanks for the tip.
```Julia
julia> @code_warntype c[1:2, :a]
Variables
  #self#::Core.Compiler.Const(getindex, false)
  A::DenseAxisArray{Int64,2,Tuple{Base.OneTo{Int64},Array{Symbol,1}},Tuple{JuMP.Containers._AxisLookup{Base.OneTo{Int64}},JuMP.Containers._AxisLookup{Dict{Symbol,Int64}}}}
  idx::Tuple{UnitRange{Int64},Symbol}
  new_indices::Tuple{Array{Int64,1},Int64}
  new_axes::Tuple{Array{Int64,1}}

Body::DenseAxisArray{Int64,1,Tuple{Array{Int64,1}},Tuple{JuMP.Containers._AxisLookup{Dict{Int64,Int64}}}}
1 ─       Core.NewvarNode(:(new_axes))
│   %2  = Base.to_index::Core.Compiler.Const(Base.to_index, false)
│         (new_indices = (%2)(A, idx))
│   %4  = JuMP.Containers.any(JuMP.Containers._is_range, new_indices)::Bool
│   %5  = !%4::Bool
└──       goto #3 if not %5
2 ─ %7  = Base.getproperty(A, :data)::Array{Int64,2}
│   %8  = Core.tuple(%7)::Tuple{Array{Int64,2}}
│   %9  = Core._apply_iterate(Base.iterate, Base.getindex, %8, new_indices)::Array{Int64,1}
│         Core.typeassert(%9, $(Expr(:static_parameter, 1)))
└──       Core.Compiler.Const(:(return %10), false)
3 ┄ %12 = Base.getproperty(A, :axes)::Tuple{Base.OneTo{Int64},Array{Symbol,1}}
│         (new_axes = JuMP.Containers._new_axes(%12, new_indices))
│   %14 = Base.getproperty(A, :data)::Array{Int64,2}
│   %15 = Core.tuple(%14)::Tuple{Array{Int64,2}}
│   %16 = Core._apply_iterate(Base.iterate, Base.getindex, %15, new_indices)::Array{Int64,1}
│   %17 = Core.tuple(%16)::Tuple{Array{Int64,1}}
│   %18 = Core._apply_iterate(Base.iterate, JuMP.Containers.DenseAxisArray, %17, new_axes)::DenseAxisArray{Int64,1,Tuple{Array{Int64,1}},Tuple{JuMP.Containers._AxisLookup{Dict{Int64,Int64}}}}
└──       return %18
```

Reply:
Sweet